guibot/path.py

```diff
@@ -113,3 +113,47 @@ def search(self, filename, restriction="", silent=False):
             raise FileNotFoundError('File ' + filename + ' not found')
 
         return None
+
+    def __iter__(self):
+        for p in self._target_paths:
+            yield p
+
+    def __len__(self):
+        return len(self._target_paths)
+
+
+class ScopedPath(object):
+    """
+    Class to be used to search for files
+    inside certain paths.
+    """
+    def __init__(self, *paths):
+        """
+        Create the class with the paths that the search will be
+        restricted to.
+
+        :param paths: list of paths that the search will use
+        """
+        self._paths = paths
+
+    def __enter__(self):
+        """
+        Enter the restricted context, where any search will be
+        restricted to the paths used to initialize this class.
+        """
+        path_searcher = Path()
+        self._old_paths = list(path_searcher)
+        path_searcher.clear()
+        for p in self._paths:
+            path_searcher.add_path(p)
+        return path_searcher
```

Comment:
Should the old paths be a superset of the new paths? This is what I understand with scoping, as in paths within the previous paths. Here it seems to me that I can just provide any paths (not necessarily a subset) and replace the ones of the path searcher independently.

Reply:
No, "scope" refers to the extent of the search, meaning the search will be restricted to that paths only. This superset concept is already what `Path` uses: it searches paths in the order they are added, meaning that paths added later would be a superset of the ones added before.

Reply:
I am not sure how the order of paths here defines a superset relation but I guess what you meant above is simply "temporarily replaced the set of currently added paths with a set of different ones". There is no relation between the first and second sets of paths at least as concluded from this implementation. I took the word "scope" here for a bit misleading but I guess it could stay here in the sense of scope of a context manager.

Could we maybe change the class docstring a bit to clarify this point? The string could say something like "temporarily replace the currently loaded paths" etc. etc. I guess `TemporaryPath` is not better for a class name?

Reply:
I've extended the docstring to clarify this a bit.
I think that `Path` here is a bit misleading as there's no operation on paths (you even confused the meaning of `Path().remove_path` in another PR). So I'd say "Path" should be replaced by "PathSearcher" or maybe "FileSearcher". Then one would be "FileSearcher" and the other "ScopedFileSearcher", which would be closer to the behavior of the class.

Reply:
Yes, I have also found the usage of the `Path` name confusing in the past and it predates most of my changes too. I am wondering if `Searcher` is a good name though as we already have `Finder` and it becomes a bit overlapping. Maybe `PathChecker`? In any case we can think about this in the future too. What do you think?

Reply:
Sure, agreed

Reply:
I recently had to use the `TemporaryConfig` class as a context manager to get some temporary configuration where I needed it. Using the same logic, could this be called `TemporaryPath`? From what I have gotten so far this is exactly what it does - temporarily replaces the paths of that are already set and restores them once done.

Reply:
Consistency-wise I agree. However, "TemporaryConfig" has a clear meaning: it's a temporary configuration. "TemporaryPath", on the other hand, doesn't mean what it seems. To me it looks like a single temporary path is being created somewhere and returned, similarly to the built-in [TemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile). And the fact that `TemporaryFile` can also be used as a context manager makes the name "TemporaryPath" even more confusing:
```python
from tempfile import TemporaryFile
with TemporaryFile() as fp:
     fp.write(b'Hello world!')
     fp.seek(0)
     fp.read()
```

Of course, most of the confusion yields from the broad name of the "Path" class. I'm still in for giving it a proper name, because I think that adding the prefix "Temporary" in front of an already confusing name will make things even worse.

Consistency-wise again, the name "Path" is completely different from the rest of the other names. For example, "Target" and "Location" respectively represent an object, in this case a target and a location, while "Path" doesn't represent a "Path". Additionally, "Find**er**" and "ImageLogg**er**" are like actors in that they  take action, i.e. they respectively _find_ and _log_ something, but even though "Path" _searches_ for something, it doesn't end with **er**.

With that said, I personally favor the alternatives:
- `class PathResolver:` / `class FileResolver:` / `class Resolver:` (or "Locator" instead of "Resolver")
This has a clearer meaning as the class will be turning an incomplete/relative name into an accessible file. This is consistent with `Finder` and `ImageLogger`.

```python
FileResolver().add_path("/some/path")
filename = FileResolver().search(filename)

# "Scoped" (or "Restricted") because the "resolving" operation is scoped
with ScopedFileResolver("/search/only/here"):
    filename = FileResolver().search(filename)
```

- `class Paths:` / `class PathList`
This is another alternative as the "Path" class is simply a collection of paths with a convenience "search" method.  This is consistent with `Target` and `Location`.

```python
# simply "append" - it's a list of paths so it's clear what we are appending
PathList().append("/some/path")
# "search_file" because we're searching a file, not a path from the list
filename = PathList().search_file(filename)

# "Temporary" because the object (the path list) is temporary
with TemporaryPathList("/search/only/here"):
    filename = PathList().search_file(filename)
```

Reply:
> Consistency-wise I agree. However, "TemporaryConfig" has a clear meaning: it's a temporary configuration. "TemporaryPath", on the other hand, doesn't mean what it seems. To me it looks like a single temporary path is being created somewhere and returned, similarly to the built-in [TemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile). And the fact that `TemporaryFile` can also be used as a context manager makes the name "TemporaryPath" even more confusing:
> 
> ```python
> from tempfile import TemporaryFile
> with TemporaryFile() as fp:
>      fp.write(b'Hello world!')
>      fp.seek(0)
>      fp.read()
> ```

I guess this is a valid argument but the fact that some class is used in some way shouldn't immediately suggest the way another class is used.

> Of course, most of the confusion yields from the broad name of the "Path" class. I'm still in for giving it a proper name, because I think that adding the prefix "Temporary" in front of an already confusing name will make things even worse.
> 
> Consistency-wise again, the name "Path" is completely different from the rest of the other names. For example, "Target" and "Location" respectively represent an object, in this case a target and a location, while "Path" doesn't represent a "Path". Additionally, "Find**er**" and "ImageLogg**er**" are like actors in that they take action, i.e. they respectively _find_ and _log_ something, but even though "Path" _searches_ for something, it doesn't end with **er**.

Completely agree, the name made sense in the past but is not too outdated to bear its original meaning.

> With that said, I personally favor the alternatives:
> 
>     * `class PathResolver:` / `class FileResolver:` / `class Resolver:` (or "Locator" instead of "Resolver")
>       This has a clearer meaning as the class will be turning an incomplete/relative name into an accessible file. This is consistent with `Finder` and `ImageLogger`.

I think `PathResolver` or `FileResolver` makes most practical sense similarly to the `ImageLogger` since simply put we just resolve file paths used for tagets, match configurations, etc. What do you think? Should we also consult with @thomasjfox who likes chiming in when it comes to renames?

I am having some thoughts regarding stability since it is possible many users already imported and use `Path` in their scripts so perhaps at least for a while we could deprecate the class (with a warning) and move everything to the new module. We are still in the stage of unstable API so we didn't promise anything to anyone but at least as a minimal action of courtesy. Opinions here?

> ```python
> FileResolver().add_path("/some/path")
> filename = FileResolver().search(filename)
> 
> # "Scoped" (or "Restricted") because the "resolving" operation is scoped
> with ScopedFileResolver("/search/only/here"):
>     filename = FileResolver().search(filename)
> ```

I am still confused by the usage of "Scoped" and "Restricted" here which suggests subsets, i.e. a selection of paths from the original paths, while the actual implementation *replaces* the paths with arbitrary paths. Am I missing something vital here?

>     * `class Paths:` / `class PathList`
>       This is another alternative as the "Path" class is simply a collection of paths with a convenience "search" method.  This is consistent with `Target` and `Location`.
> 
> 
> ```python
> # simply "append" - it's a list of paths so it's clear what we are appending
> PathList().append("/some/path")
> # "search_file" because we're searching a file, not a path from the list
> filename = PathList().search_file(filename)
> 
> # "Temporary" because the object (the path list) is temporary
> with TemporaryPathList("/search/only/here"):
>     filename = PathList().search_file(filename)
> ```

Yes but maybe we should go for a more descriptive and unambiguous name this time, especially since "path" has wider and more varied uses.

So I guess this is a fairly small problem that boils down to the choice of name. Do you think "path" or "file" relates better to what the module is doing as a term?

Reply:
"scoped" and "restricted" indeed sounds a bit like
it's just a subset of the original path list.

How about "Custom" prefix? Something like:

```
with CustomPathResolver("/search/only/here"):
    filename = FileResolver().search(filename)

with CustomSearchPath("/search/only/here"):
    filename = FileResolver().search(filename)

with OverrideSearchPath("/search/only/here"):
    filename = FileResolver().search(filename)
```
Personally I like "CustomSearchPath" best.

Reply:
@pevogam 
> I am having some thoughts regarding stability since it is possible many users already imported and use Path in their scripts so perhaps at least for a while we could deprecate the class (with a warning) and move everything to the new module. We are still in the stage of unstable API so we didn't promise anything to anyone but at least as a minimal action of courtesy. Opinions here?

I agree, we should keep the original class for compatibility for at least a couple of minors.

> I am still confused by the usage of "Scoped" and "Restricted" here which suggests subsets, i.e. a selection of paths from the original paths, while the actual implementation replaces the paths with arbitrary paths. Am I missing something vital here?

Indeed. It was an attempt to convey the fact that an alternative list of paths will be used, that the search will be "restricted" to another collection of paths. In this regard, I think Tom nailed it with "Custom".

> Do you think "path" or "file" relates better to what the module is doing as a term?

Personally I would go for "file", but I don't dislike "path" either. Your call :slightly_smiling_face:

Reply:
I think let's conclude this with `FileResolver` and `CustomFileResolver`, what do you think?

Reply:
@pevogam agreed and adjusted.